src/Nerdbank.Streams.Tests/NestedStreamTests.cs

```diff
@@ -90,13 +100,79 @@ public void SetLength()
+
+        Assert.Equal(5, this.stream.Seek(5, SeekOrigin.Current));
+        Assert.Equal(5, this.underlyingStream.Position);
+        Assert.Equal(5, this.stream.Seek(0, SeekOrigin.Current));
+        Assert.Equal(5, this.underlyingStream.Position);
+        Assert.Equal(4, this.stream.Seek(-1, SeekOrigin.Current));
+        Assert.Equal(4, this.underlyingStream.Position);
+        Assert.Equal(0, this.stream.Seek(-10, SeekOrigin.Current));
+        Assert.Equal(0, this.underlyingStream.Position);
+
+        Assert.Equal(DefaultNestedLength, this.stream.Seek(DefaultNestedLength + 1, SeekOrigin.Current));
+        Assert.Equal(DefaultNestedLength, this.underlyingStream.Position);
+        Assert.Equal(DefaultNestedLength, this.stream.Seek(DefaultNestedLength, SeekOrigin.Current));
+        Assert.Equal(DefaultNestedLength, this.underlyingStream.Position);
+        Assert.Equal(DefaultNestedLength, this.stream.Seek(0, SeekOrigin.Current));
+        Assert.Equal(DefaultNestedLength, this.underlyingStream.Position);
+        Assert.Equal(0, this.stream.Seek(-20, SeekOrigin.Current));
+        Assert.Equal(0, this.underlyingStream.Position);
+
+        this.stream.Dispose();
+        Assert.Throws<ObjectDisposedException>(() => this.stream.Seek(0, SeekOrigin.Begin));
+    }
+
+    [Fact]
+    public void Seek_Begin()
     {
-        Assert.Throws<NotSupportedException>(() => this.stream.Seek(0, SeekOrigin.Begin));
+        Assert.Equal(0, this.stream.Position);
+        Assert.Equal(0, this.stream.Seek(-1, SeekOrigin.Begin));
+        Assert.Equal(0, this.underlyingStream.Position);
+
+        Assert.Equal(0, this.stream.Seek(0, SeekOrigin.Begin));
+        Assert.Equal(0, this.underlyingStream.Position);
+
+        Assert.Equal(5, this.stream.Seek(5, SeekOrigin.Begin));
+        Assert.Equal(5, this.underlyingStream.Position);
+
+        Assert.Equal(DefaultNestedLength, this.stream.Seek(DefaultNestedLength, SeekOrigin.Begin));
+        Assert.Equal(DefaultNestedLength, this.underlyingStream.Position);
+
+        Assert.Equal(DefaultNestedLength, this.stream.Seek(DefaultNestedLength + 1, SeekOrigin.Begin));
+        Assert.Equal(DefaultNestedLength, this.underlyingStream.Position);
+
         this.stream.Dispose();
         Assert.Throws<ObjectDisposedException>(() => this.stream.Seek(0, SeekOrigin.Begin));
     }
 
+    [Fact]
+    public void Seek_End()
+    {
+        Assert.Equal(0, this.stream.Position);
+        Assert.Equal(9, this.stream.Seek(-1, SeekOrigin.End));
+        Assert.Equal(9, this.underlyingStream.Position);
+
+        Assert.Equal(DefaultNestedLength, this.stream.Seek(0, SeekOrigin.End));
+        Assert.Equal(DefaultNestedLength, this.underlyingStream.Position);
+
+        Assert.Equal(DefaultNestedLength, this.stream.Seek(5, SeekOrigin.End));
+        Assert.Equal(DefaultNestedLength, this.underlyingStream.Position);
+
+        Assert.Equal(0, this.stream.Seek(-20, SeekOrigin.Begin));
```

Comment:
`MemoryStream` and `FileStream` throw when seeking to a position before the first position. We should do the same.